### ISSUE TITLE
kind of TopicPartitionBlock with maps

### ISSUE THREAD

**Issue by DEvil0000** [danieljoos/libkafka-asio#12]
For requests it makes perfect sense to use `vectors` to store the information because you normally just add data and don't search it.
But for responses it would be useful to have the `TopicPartitionBlocks` as `map` structure (because you will most likely search it).
With the functions in `detail/functional` you build some kind of simple data search and access functions. If the structure would be a map the access would be faster and simple. And we would be able to just use it in the `Client` or where ever without the need of reorganising data.

Do you agree?

edit:
why are you using the `OptionalType` there?
```
typedef boost::optional<Partition> OptionalType;
```
sub classing the struct should be enough right?
```
struct Partition :
    public TPartitionProperties
  {
  }
```

edit2:
may not be true for all response types but for meta data request I think it is

**Comment by DEvil0000** [comment]
I refactored this in a branch so you can test it.
I am not sure if this makes sense for all responses or just some of them so I commited each one separate. Feel free to cherry-pick or tell me which commits you like to merge - then I can sort the unwanted ones out ;)
https://github.com/DEvil0000/libkafka-asio/tree/TopicPartitionMap

**Comment by danieljoos** [comment]
The `Topic::Partition::OptionalType` is used e.g. by the `OffsetResponse::TopicPartitionOffset` function as return value:
https://github.com/danieljoos/libkafka-asio/blob/master/examples/offset_cxx03.cpp#L35
The function expects a topic name and a partition number. If there is no offset information for that topic-partition available inside the `OffsetResponse` object, the result will be empty. It's just a type definition anyway, so no overhead regarding the size of the struct.

**Comment by danieljoos** [comment]
Using a map sounds good to me. Your code also looks very good. Could you please adapt the naming of the variables (I'm trying to follow this guideline, using underscores instead of mixed-case: https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Variable_Names) to be consistent with the rest of the lib? Create a pull request and we get that stuff into the lib :smile:

**Comment by DEvil0000** [comment]
In fact you just use a subset of this coding guideline ;)
for example you place the brackets in a new line while following the google guide they have to be in the same line.
I will try to follow your style but it is hard for me - every project has a different coding style -.-
Most projects I am working on use mixed case variable names and have 4 space indent.

**Comment by DEvil0000** [comment]
or do you want a request to the feature branch?

**Comment by danieljoos** [comment]
Fixed with #13
